[SPARK-9862]Handling data skew

YuhuWang2002 opened the pull request:
## What changes were proposed in this pull request?

   Generally reducer need fetch all mapper block to deal。But in data skew case，huge data in one reducer, make the running time is too long, or Out Of Memory.  So we need design a new feature to handling skew data join.  Since a reduce to deal with the data is too large, then scattered to a number of reducer for processing.
In join case. A join B equal to A1 join B + A2 join B + …. An join B.  So when in join case the data skew occur. Don’t pull all skew data into a reducer do as A join B, but pull one mapper data and whole data about other join side do  as A1 join B + A2 join B + …. An join B.

In spark sql, we use ShuffedRowRDD to fetch the map side generate date. ShuffedRowRDD only can read  a range of reduce IDs but fetches from all maps. Join operater generate two ShuffedRowRDD, them has the same partition num, and put the same partition id together to join. The ShuffedRowRDD partition  like this:
ShuffledRowRDD 1: [1,3,5,9]
ShuffledRowRDD 2: [1,3,5,9]

For handling the data skew case. We design a new RDD to read one mapper , we call it SkewShuffleRowRDD. The SkewShuffleRowRDD function getPartitions can generate partition to describe read one or a range of mappers data or read all mappers data.
The the two SkewShuffleRowRDD partition like this:
SkewShuffleRowRDD 1：
[(1, read all maps),(3, read map 1 to 3),(3,read map 3 to 5),(3,read map 5 to 7),(5, read all maps),(9,read all maps)]
SkewShuffleRowRDD 2：
[(1, read all maps),(3, read all maps),(3, read all maps),(3, read all maps),(5, read all maps),(9,read all maps)]

For determine which partition is data skew,  add a new configure parameter : spark.sql.adaptive.skewjoin.threshold.  when the skew threshold  > map out size. Determine the partition is data skew.
For achieve this feature, need add some code in ExchangeCoordinator.scala  to aware the skew partition. If is join operater and find some partition is skew, then generate the SkewShuffleRowRDD instead ShuffleRowRDD
## How was this patch tested?

Unit tests in ExchangeCoordinatorSuite

also can generate skew data and  manual test

Author: wangyuhuwangyuhu2002@126.com

Commit:
[SPARK-9862] handling data skew , add skew join feature

Commit:
code style fix

Commit:
add comment

Commit:
modify UT code

AmplabJenkins commented:
Can one of the admins verify this patch?

hvanhovell commented:
test this please

SparkQA commented:
**[Test build #66106 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66106/consoleFull)** for PR 15297 at commit [`0ba86a2`](https://github.com/apache/spark/commit/0ba86a2284e684a733f989ed6e595575f511c8bd).

SparkQA commented:
**[Test build #66106 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66106/consoleFull)** for PR 15297 at commit [`0ba86a2`](https://github.com/apache/spark/commit/0ba86a2284e684a733f989ed6e595575f511c8bd).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66106/
Test FAILed.

scwf commented:
retest this please.

scwf commented:
retest this please

hvanhovell commented:
ok to test

SparkQA commented:
**[Test build #66564 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66564/consoleFull)** for PR 15297 at commit [`5786e22`](https://github.com/apache/spark/commit/5786e22132c9e80f03ce45e928d1f2de962cdd16).

SparkQA commented:
**[Test build #66564 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66564/consoleFull)** for PR 15297 at commit [`5786e22`](https://github.com/apache/spark/commit/5786e22132c9e80f03ce45e928d1f2de962cdd16).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66564/
Test FAILed.

Commit:
fix scala style

SparkQA commented:
**[Test build #66569 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66569/consoleFull)** for PR 15297 at commit [`4586ed6`](https://github.com/apache/spark/commit/4586ed6c7b271fc1df92a77635b3f0b61242630f).

YuhuWang2002 commented:
test this please

wzhfy commented:
retest this please

SparkQA commented:
**[Test build #66570 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66570/consoleFull)** for PR 15297 at commit [`4586ed6`](https://github.com/apache/spark/commit/4586ed6c7b271fc1df92a77635b3f0b61242630f).

wzhfy commented:
Please describe the pr more clearly instead of using links.

SparkQA commented:
**[Test build #66570 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66570/consoleFull)** for PR 15297 at commit [`4586ed6`](https://github.com/apache/spark/commit/4586ed6c7b271fc1df92a77635b3f0b61242630f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66570/
Test FAILed.

SparkQA commented:
**[Test build #66569 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66569/consoleFull)** for PR 15297 at commit [`4586ed6`](https://github.com/apache/spark/commit/4586ed6c7b271fc1df92a77635b3f0b61242630f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66569/
Test PASSed.

witgo commented on sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ExchangeCoordinator.scala:
```diff
@@ -183,13 +186,142 @@ class ExchangeCoordinator(
+    partitionStartIndices: Option[Array[Int]] = None) : Array[(Int, Array[(Int, Int, Int)])] = {
+    // find which partition is skew
+    var skewPartition = mapOutputStatistics.map(ms =>
+      ms.bytesByPartitionId.zipWithIndex
```
` ms.bytesByPartitionId.zipWithIndex.view`

Commit:
add commet and refactor code

Commit:
remove unnecessary and add comment

SparkQA commented:
**[Test build #66647 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66647/consoleFull)** for PR 15297 at commit [`ac66284`](https://github.com/apache/spark/commit/ac66284de05493783c787db752c17a2db9d79ea0).

SparkQA commented:
**[Test build #66647 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66647/consoleFull)** for PR 15297 at commit [`ac66284`](https://github.com/apache/spark/commit/ac66284de05493783c787db752c17a2db9d79ea0).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66647/
Test FAILed.

YuhuWang2002 commented:
retest this please

wzhfy commented:
retest this please

SparkQA commented:
**[Test build #66648 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66648/consoleFull)** for PR 15297 at commit [`ac66284`](https://github.com/apache/spark/commit/ac66284de05493783c787db752c17a2db9d79ea0).

SparkQA commented:
**[Test build #66648 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66648/consoleFull)** for PR 15297 at commit [`ac66284`](https://github.com/apache/spark/commit/ac66284de05493783c787db752c17a2db9d79ea0).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66648/
Test FAILed.

wzhfy commented:
retest this please

SparkQA commented:
**[Test build #66650 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66650/consoleFull)** for PR 15297 at commit [`ac66284`](https://github.com/apache/spark/commit/ac66284de05493783c787db752c17a2db9d79ea0).

SparkQA commented:
**[Test build #66650 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/66650/consoleFull)** for PR 15297 at commit [`ac66284`](https://github.com/apache/spark/commit/ac66284de05493783c787db752c17a2db9d79ea0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/66650/
Test PASSed.

SaintBacchus commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SkewShuffleRowRDD.scala:
```diff
@@ -0,0 +1,147 @@
+import org.apache.spark.sql.catalyst.InternalRow
+
+class SkewCoalescedPartitioner(
+        val parent: Partitioner,
```
Nit: code format

SaintBacchus commented on core/src/main/scala/org/apache/spark/MapOutputTracker.scala:
```diff
@@ -138,13 +138,16 @@ private[spark] abstract class MapOutputTracker(conf: SparkConf) extends Logging
    */
-  def getMapSizesByExecutorId(shuffleId: Int, startPartition: Int, endPartition: Int)
+  def getMapSizesByExecutorId(shuffleId: Int, startPartition: Int, endPartition: Int,
+                              mapid: Int = -1)
```
It's better to use `Seq[Int]` to fetch many maps in one time.

witgo replied:
`mapid` => `mapId`

YuhuWang2002 replied:
it's a good idea. the seq[Int] parameter can fetch more maps data. it can reduce the task num

witgo commented on core/src/main/scala/org/apache/spark/MapOutputTracker.scala:
```diff
@@ -687,18 +691,21 @@ private[spark] object MapOutputTracker extends Logging {
       endPartition: Int,
-      statuses: Array[MapStatus]): Seq[(BlockManagerId, Seq[(BlockId, Long)])] = {
+      statuses: Array[MapStatus],
+      mapIdx: Int = -1): Seq[(BlockManagerId, Seq[(BlockId, Long)])] = {
```
How to change `mapIdx` to b`mapId` ?

YuhuWang2002 replied:
it's conflicts with mapId

witgo commented on core/src/main/scala/org/apache/spark/shuffle/BlockStoreShuffleReader.scala:
```diff
@@ -35,7 +35,8 @@ private[spark] class BlockStoreShuffleReader[K, C](
     blockManager: BlockManager = SparkEnv.get.blockManager,
-    mapOutputTracker: MapOutputTracker = SparkEnv.get.mapOutputTracker)
+    mapOutputTracker: MapOutputTracker = SparkEnv.get.mapOutputTracker,
+    mapid: Int = -1)
```
`mapid` => `mapId`

witgo commented on core/src/main/scala/org/apache/spark/shuffle/ShuffleManager.scala:
```diff
@@ -48,7 +48,8 @@ private[spark] trait ShuffleManager {
       endPartition: Int,
-      context: TaskContext): ShuffleReader[K, C]
+      context: TaskContext,
+      mapid: Int = -1): ShuffleReader[K, C]
```
`mapid` => `mapId`

witgo commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SkewShuffleRowRDD.scala:
```diff
@@ -0,0 +1,147 @@
+  override def getPartitions: Array[Partition] = {
+    val partitions = ArrayBuffer[Partition]()
+    var partitionIndex = -1
+    for(i <- 0 until partitionStartIndices.length ) {
```
` for(i <- 0 until partitionStartIndices.length )` -> `  for (i <- partitionStartIndices.indices) `

witgo commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SkewShuffleRowRDD.scala:
```diff
@@ -0,0 +1,147 @@
+  @transient private lazy val parentPartitionMapping: Array[Int] = {
+    val n = parent.numPartitions
+    val result = new Array[Int](n)
+    for (i <- 0 until partitionStartIndices.length) {
```
`for (i <- 0 until partitionStartIndices.length) ` ->`for (i <- partitionStartIndices.indices) `

ron8hu commented:
cc @rxin @hvanhovell  Can you review and comment this PR?  Thanks.

ron8hu commented:
The design note of this PR has been posted at jira page.

Commit:
read a range of maps

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

YuhuWang2002 commented:
I will rewrite the read ShuffleReader  interface , for read a range of maps but not only read a map data.
it will be finished soon

Commit:
refactor code for easy read

Commit:
read a range of maps

SparkQA commented:
**[Test build #67162 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67162/consoleFull)** for PR 15297 at commit [`36624ec`](https://github.com/apache/spark/commit/36624ec94fdc3a2eeeadea1d86cdf2e794198434).

YuhuWang2002 commented:
following the review comment， I rewrite code for read a range of maps。like this：
class BlockStoreShuffleReader[K, C](
    handle: BaseShuffleHandle[K, _, C],
    startPartition: Int,
    endPartition: Int,
    context: TaskContext,
    serializerManager: SerializerManager = SparkEnv.get.serializerManager,
    blockManager: BlockManager = SparkEnv.get.blockManager,
    mapOutputTracker: MapOutputTracker = SparkEnv.get.mapOutputTracker,
    startMapId: Option[Int] = None,
    endMapId: Option[Int] = None)

To decide how many range for read from the maps。Use the spark.sql.adaptive.skewjoin.threshold value。We think the output size less than the skew threshold， It can handling in one task，else we split to many task，which every one task handing data size slightly less the skew threshold

SaintBacchus commented on core/src/main/scala/org/apache/spark/MapOutputTracker.scala:
```diff
@@ -138,13 +138,17 @@ private[spark] abstract class MapOutputTracker(conf: SparkConf) extends Logging
    */
-  def getMapSizesByExecutorId(shuffleId: Int, startPartition: Int, endPartition: Int)
+  def getMapSizesByExecutorId(shuffleId: Int, startPartition: Int, endPartition: Int,
+                              startMapId: Option[Int] = None,
```
If use `startMapId` and `endMapId`, it's too difficulty for the partition to be aware of the localtion. I suggest use `Seq[Int]` instead. Since the `Seq[Int]` will have a big size in closure, you can `bitmap` also.

SparkQA commented:
**[Test build #67162 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67162/consoleFull)** for PR 15297 at commit [`36624ec`](https://github.com/apache/spark/commit/36624ec94fdc3a2eeeadea1d86cdf2e794198434).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67162/
Test FAILed.

Commit:
refactor code

SparkQA commented:
**[Test build #67175 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67175/consoleFull)** for PR 15297 at commit [`3682a9e`](https://github.com/apache/spark/commit/3682a9ecf72caebc186cc1539dfa0ebe8683aef2).

SparkQA commented:
**[Test build #67175 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67175/consoleFull)** for PR 15297 at commit [`3682a9e`](https://github.com/apache/spark/commit/3682a9ecf72caebc186cc1539dfa0ebe8683aef2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67175/
Test FAILed.

Commit:
fix UT fails

SparkQA commented:
**[Test build #67243 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67243/consoleFull)** for PR 15297 at commit [`6c943d1`](https://github.com/apache/spark/commit/6c943d18f4cde1b52a6c26731d7bc18789743942).

SparkQA commented:
**[Test build #67243 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67243/consoleFull)** for PR 15297 at commit [`6c943d1`](https://github.com/apache/spark/commit/6c943d18f4cde1b52a6c26731d7bc18789743942).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67243/
Test PASSed.

tgravescs commented:
I haven't looked through the code in detail but can you clarify the design a bit on this, the design pretty much just says we are splitting up the fetch of the map outputs but it doesn't say what happens then or how this really solves the problem.  

You say that instead of doing A join B you are splitting it up to do something like A1 join B + A2 join B + …. An join B.  Is it still just one reduce task fetching it in separate chunks if so how does this fix the problem or is it treating each one of those fetches as a separate partitions when processing it?

YuhuWang2002 commented:
@tgravescs ：
Thank you for your response, when a single reduce task handling huge data, it's slowly and unstable. so we split one reduce task to multi- reduce task.
A single reduce task doing like A join B. we split to multi-task. task 1 doing A1 join B,  task 2 dong A2 join B and so on.  A1 is a part of A which read from a range of maps output.  For spark sql, it is the A1 as a  separate partitions when processing. so it can use mutil-executor to run the task.  for dispersion the process pressure.

tgravescs commented:
Ok so how does that affect the overall job and # of outputs?  I don't know the internals of Spark SQL so sorry if I'm missing something obvious.  Basically now you will have multiple tasks whereas it used to use 1.  So lets say I have spark.sql.shuffle.partitions=200 to start with, the skewed join add tasks to process some skewed partition so lets say it runs 210 tasks, then lets say I save that to HDFS, do I get 210 output files or does it join those 10 back into 1 again?

YuhuWang2002 commented:
@tgravescs : In join case,some like : select count(*) from A join B.  if the parameter spark.sql.shuffle.partitions=200 ,then we get 200 tasks output about 'count num', the output is not in HDFS but cache in spark . Calculate the sum of 200 tasks. we got the correct value.  If skewed. wo get 210 tasks  output about 'count num'.  it's some processing about next step.

tgravescs commented:
Ok so are you saying this skewed join implementation doesn't apply to other dataframe operations, something like:

 val df_pixels = sqlContext.read.parquet("somefile")
    val df_pixels_renamed = df_pixels.withColumnRenamed("photo_id", "pixels_photo_id")
    val df_meta = sqlContext.read.parquet("somemeta")
    val df = df_meta.as("meta").join(df_pixels_renamed, $"meta.photo_id" === $"pixels_photo_id", "inner").drop("pixels_photo_id")
    df.write.parquet("someoutputfile")

Where normally spark.sql.shuffle.partitions=X would configure the number of output files.  So in my example if I set spark.sql.shuffle.partitions=200 but skewed join use 210, what happens, how many output files would I get?

YuhuWang2002 commented:
skewed join implementation suit for dataframe and sql statement
you will get 210 output files.

YuhuWang2002 commented:
I do some performance test between use skew join algorithm and not use skew join  algorithm.
I generate 2 table with 1/5 data skew in table S and 1/10000 data skew in table R. Two table skew in same key.

spark.sql.adaptive.skewjoin.threshold   6000000
spark.sql.adaptive.shuffle.targetPostShuffleInputSize   5000000
record: S 10000000 rows; R 100000000 rows
sql:
select count(*) from R,S where rid=sid and sname>'wang9' and rname > 'zhang9';

skew algorithm : 167.695s
normal algorithm: 303.922s

R2_txt is 100000000 rows without data skew.
sql: select count(*) from R2_txt,S where rid=sid and sname>'wang' and rname > 'zhang9';
skew algorithm : 38.717s
normal algorithm: 114.21s

scwf commented:
@YuhuWang2002 
We should limit the use case for outer join: 
For left outer join, such as A left join B, this implementation now can not handle the case of  skew of  table B.  That's because the result of join depends on the all data of the same reduce data of B, you can not split it to multi-tasks.

Similarly, for right outer join, such as A right join B, this implementation now can not handle the case of  skew of  table A. And for full outer join, we can not use the optimization.

Commit:
limit the use case for outer join

Commit:
add ut for out join

SparkQA commented:
**[Test build #68442 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68442/consoleFull)** for PR 15297 at commit [`b60f9bc`](https://github.com/apache/spark/commit/b60f9bc76763a0c149cb32bf8b3ab3f318a86635).

SparkQA commented:
**[Test build #68442 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68442/consoleFull)** for PR 15297 at commit [`b60f9bc`](https://github.com/apache/spark/commit/b60f9bc76763a0c149cb32bf8b3ab3f318a86635).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68442/
Test FAILed.

SparkQA commented:
**[Test build #68446 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68446/consoleFull)** for PR 15297 at commit [`8728d33`](https://github.com/apache/spark/commit/8728d334a79f3cb385937d61956ea47d2e9a4650).

SparkQA commented:
**[Test build #68446 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68446/consoleFull)** for PR 15297 at commit [`8728d33`](https://github.com/apache/spark/commit/8728d334a79f3cb385937d61956ea47d2e9a4650).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68446/
Test FAILed.

Commit:
fix style error

SparkQA commented:
**[Test build #68448 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68448/consoleFull)** for PR 15297 at commit [`1bb158b`](https://github.com/apache/spark/commit/1bb158b3035cd4f69dd2f47c26ef1c67bc5e6a6c).

SparkQA commented:
**[Test build #68448 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68448/consoleFull)** for PR 15297 at commit [`1bb158b`](https://github.com/apache/spark/commit/1bb158b3035cd4f69dd2f47c26ef1c67bc5e6a6c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68448/
Test PASSed.

holdenk commented:
This is a really big change - and handling skewed data in joins is certainly an important consideration - have you considered making a design document and running it by the dev list? Maybe something similar to the recently proposed Spark Improvement Proposals process?

SparkQA commented:
**[Test build #69990 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69990/consoleFull)** for PR 15297 at commit [`99b8305`](https://github.com/apache/spark/commit/99b830584aafb53112b5bdd2d723080fa19baa54).

SparkQA commented:
**[Test build #69990 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/69990/consoleFull)** for PR 15297 at commit [`99b8305`](https://github.com/apache/spark/commit/99b830584aafb53112b5bdd2d723080fa19baa54).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/69990/
Test FAILed.

YuhuWang2002 commented:
retest this please

deswal-ajit commented:
hi

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

albertusk95 commented on sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ExchangeCoordinator.scala:
```diff
@@ -184,10 +189,146 @@ class ExchangeCoordinator(
+     return (0 until numExchanges).map(_ =>
+       (0, Array[(Int, Long, Int, Int)]((-1, 0, 0, 0)))).toArray
+   }
+   // skewSize must great than TargetPostShuffleInputSize
```
`// skewSize must greater than TargetPostShuffleInputSize`

AmplabJenkins commented:
Can one of the admins verify this patch?
